packages/flutter_tools/lib/src/ios/core_devices.dart

```diff
@@ -10,9 +10,225 @@ import '../base/file_system.dart';
+    }
+
+    final IOSCoreDeviceRunningProcess? launchedProcess = launchResult.process;
+    final int? processId = launchedProcess?.processIdentifier;
+    if (launchedProcess == null || processId == null) {
+      return false;
+    }
+
+    // Start LLDB and attach to the device process.
+    final bool attachStatus = await _lldb.attachAndStart(deviceId, processId);
+
+    // If it fails to attach with lldb, kill the launched process so it doesn't stay hanging.
+    if (!attachStatus) {
+      await stopApp(deviceId: deviceId, processId: processId);
+      return false;
+    }
+    return attachStatus;
+  }
+
+  /// Install and launch the app on the device through Xcode using Mac Automation ([_xcodeDebug]).
+  Future<bool> launchAppWithXcodeDebugger({
+    required String deviceId,
+    required DebuggingOptions debuggingOptions,
+    required IOSApp package,
+    required List<String> launchArguments,
+    required TemplateRenderer templateRenderer,
+    String? mainPath,
+    @visibleForTesting Duration? discoveryTimeout,
+  }) async {
+    XcodeDebugProject? debugProject;
+
+    if (package is PrebuiltIOSApp) {
+      debugProject = await _xcodeDebug.createXcodeProjectWithCustomBundle(
+        package.deviceBundlePath,
+        templateRenderer: templateRenderer,
+        verboseLogging: _logger.isVerbose,
+      );
+    } else if (package is BuildableIOSApp) {
+      final IosProject project = package.project;
+      final Directory bundle = _fileSystem.directory(package.deviceBundlePath);
+      final Directory? xcodeWorkspace = project.xcodeWorkspace;
+      if (xcodeWorkspace == null) {
+        _logger.printError('Unable to get Xcode workspace.');
+        return false;
+      }
+      final String? scheme = await project.schemeForBuildInfo(
+        debuggingOptions.buildInfo,
+        logger: _logger,
+      );
+      if (scheme == null) {
+        return false;
+      }
+      _xcodeDebug.ensureXcodeDebuggerLaunchAction(project.xcodeProjectSchemeFile(scheme: scheme));
+
+      // Before installing/launching/debugging with Xcode, update the build
+      // settings to use a custom configuration build directory so Xcode
+      // knows where to find the app bundle to launch.
+      await _xcodeDebug.updateConfigurationBuildDir(
+        project: project.parent,
+        buildInfo: debuggingOptions.buildInfo,
+        configurationBuildDir: bundle.parent.absolute.path,
+      );
+
+      debugProject = XcodeDebugProject(
+        scheme: scheme,
+        xcodeProject: project.xcodeProject,
+        xcodeWorkspace: xcodeWorkspace,
+        hostAppProjectName: project.hostAppProjectName,
+        expectedConfigurationBuildDir: bundle.parent.absolute.path,
+        verboseLogging: _logger.isVerbose,
+      );
+    } else {
+      // This should not happen. Currently, only PrebuiltIOSApp and
+      // BuildableIOSApp extend from IOSApp.
+      _logger.printError('IOSApp type ${package.runtimeType} is not recognized.');
```

Comment:
Something to think about when working in flutter_tools:
`printTrace` is more appropriate for logs that are useful for our debugging when an issue is filed. Developers see `printError` logs though, is there anything actionable they can do at this point? This log message isn't useful to them.